src/clj/fluree/db/json_ld/policy/rules.cljc

```diff
@@ -84,68 +72,106 @@
-    (if (or view? modify?)
-      {:id          id
-       :on-property on-property
-       :on-class    on-class
-       :required?   (util/get-first-value restriction const/iri-required)
-       :default?    (and (nil? on-property) (nil? on-class)) ;; with no class or property restrictions, becomes a default policy
-       :ex-message  (util/get-first-value restriction const/iri-exMessage)
-       :view?       view?
-       :modify?     modify?
-       :query       query}
-      ;; policy has incorrectly formatted view? and/or modify?
-      ;; this might allow data through that was intended to be restricted, so throw.
-      (throw (ex-info (str "Invalid policy definition. Policies must have f:action of {@id: f:view} or {@id: f:modify}. "
-                           "Policy restriction data that failed: " restriction)
-                      {:status 400
-                       :error  :db/invalid-policy})))))
-
-(defn parse-policy-rules
-  [db policy-rules]
-  (reduce
-   (fn [acc rule]
-     (let [parsed-restriction (restriction-map rule)] ;; will throw if formatting is not valid
-       (cond
-
-         (property-restriction? parsed-restriction)
-         (add-property-restriction parsed-restriction db acc)
-
-         (class-restriction? parsed-restriction)
-         (add-class-restriction parsed-restriction db acc)
-
-         (default-restriction? parsed-restriction)
-         (add-default-restriction parsed-restriction acc)
-
-         :else
-         acc)))
-   {}
-   policy-rules))
+      (if (or view? modify?)
+        (cond-> {:id          id
+                 :on-property on-property
+                 :on-class    on-class
+                 :required?   (util/get-first-value policy-doc const/iri-required)
+                 :default?    (and (nil? on-property) (nil? on-class)) ;; with no class or property restrictions, becomes a default policy-doc
+                 :ex-message  (util/get-first-value policy-doc const/iri-exMessage)
+                 :view?       view?
+                 :modify?     modify?
+                 :query       query}
+          (not-empty subject-targets)  (assoc :s-targets subject-targets)
+          (not-empty property-targets) (assoc :p-targets property-targets)
+          (not-empty object-targets)   (assoc :o-targets object-targets))
+        ;; policy-doc has incorrectly formatted view? and/or modify?
+        ;; this might allow data through that was intended to be restricted, so throw.
+        (throw (ex-info (str "Invalid policy definition. Policies must have f:action of {@id: f:view} or {@id: f:modify}. "
+                             "Policy data that failed: " policy-doc)
+                        {:status 400
+                         :error  :db/invalid-policy}))))))
+
+(defn build-wrapper
+  [db]
+  (fn [wrapper policy]
+    (cond
+      (or (:s-targets policy)
+          (:p-targets policy)
+          (:o-targets policy))
+      (add-default-restriction policy wrapper)
```

Comment:
Am I reading this correct in that these will only work if `:on-property` or `:on-class` have no policy matches? As it stands, `:on-class` would only get checked if there was no `:on-property`. Wondering if there is an execution skipping condition here, or if always run.

Reply:
If any of the `target-*` properties are present, we ignore `onClass` and `onProperty` along the same lines of how we ignore `onClass` if there's an `onProperty`.

Reply:
I think onClass goes away and this PR gives an alternate tool to accomplish this and it was never a great thing to have anyhow. 

I think however these need to be secondary to onProperty. We want policies to compose and if a global governance person wrote a policy for e.g. ex:ssn, to apply to all dbs in the company, it would be easy to subvert it with one of these policies locally.